### ISSUE TITLE
Release notes are way too small on high DPI screen

### ISSUE THREAD

**Issue by staticfinal** [vslavik/winsparkle#88]
I integrated WinSparkle into our application and it works great! Thank you for this great project.

Our application is marked as DPI aware and the WinSparkle dialogs work well on a high DPI screen. Just the release notes are displayed way too small.

The following code seems the fix this issue:

``` C++
VARIANT zoomVariant;
VariantInit(&zoomVariant);
V_VT(&zoomVariant) = VT_I4;
V_I4(&zoomVariant) = PX(100);

m_webBrowser->ExecWB((OLECMDID)63 /*OLECMDID_OPTICAL_ZOOM*/,
                             OLECMDEXECOPT_DODEFAULT,
                             &zoomVariant,
                             NULL);
```

Just add it after the WebBrowser control is created in ui.cpp.

The only problem with this solution is that the release notes' HTML needs to be adjusted to work correctly. It doesn't seem to zoom text which specifies the font size in px.

I guess the correct solution would be to implement `IDocHostUIHandler::GetHostInfo()` and return `DOCHOSTUIFLAG_DPI_AWARE` flag. I tried to implement this but didn't get far since it has been a while since I did any COM programming... ;)

**Comment by vslavik** [comment]
I don't know what's going on here, but this issue is puzzling to me: I am aware of a deficiency in HiDPI handling: the content is (pre-Windows 10, apparently) rendered blurry, as non-HiDPI apps do, but *at the correct size*. I.e. as far as I know, absent reproduction steps from you, there is no *serious* issue here — and I say that as a developer who uses WinSparkle in a HiDPI app:

![parallels picture 7](https://cloud.githubusercontent.com/assets/145881/12716500/46244370-c8e1-11e5-9200-9b7a36452551.png)
<img width="549" alt="parallels picture 5" src="https://cloud.githubusercontent.com/assets/145881/12716501/4984fa46-c8e1-11e5-9536-eebadc8865b9.png">

If you're seeing tiny text with non-crazy CSS, that would obviously be a serious HiDPI bug, but that's the thing: I expected that to happen, but never have seen it in practice. The WebBrowser control is even *documented* to use blurry upscaling for compatibility.

--

> Just the release notes are displayed way too small.

You seem to have forgotten to include some very basic and essential information, such as your Windows version, your *WinSparkle* version (I sure hope you tested with `master` first!). You also failed to include steps to reproduce, i.e. whether the behavior is observable in the included sample (it isn't) and if it isn't, _what changes to it are necessary to see it._

> Just add it after the WebBrowser control is created in ui.cpp.

The tool you're looking for is called a *Pull Request* and that's what you should do when you have code to offer. WinSparkle is useful to you, according to you, so doing a little bit more than throwing some unfinished snippets around would be a nice way to show your appreciation.

> The only problem with this solution is that the release notes' HTML needs to be adjusted to work correctly. It doesn't seem to zoom text which specifies the font size in px.

Unfortunately, though, this sounds like a rather grave error with your code and a pretty clear indication it's not a solution at all, but just a hack. `DOCHOSTUIFLAG_DPI_AWARE` is indeed the proper solution and a PR to that effect would be appreciated.

**Comment by staticfinal** [comment]
I'm using a build of the WinSparkle master branch. I'm testing on a Windows 8.1 VM on a MacBook Pro with Retina display.

Yes, the code is unfinished. I would like to provide you with a fully working patch but I'm not familiar with COM programming in wxWidgets. I just wanted to show you what I've tried so far.

Here are the release notes as they are displayed by WinSparkle:
<img width="563" alt="screen shot 2016-02-01 at 18 23 16" src="https://cloud.githubusercontent.com/assets/2419840/12725053/059ded12-c911-11e5-8db4-b1ef3069f7f6.png">

And here are the release notes as they are rendered by IE:
<img width="873" alt="screen shot 2016-02-01 at 18 23 27" src="https://cloud.githubusercontent.com/assets/2419840/12725060/2079bba2-c911-11e5-8027-859713ba5ad3.png">

Here is the HTML code:
```html
<!DOCTYPE html>
<html>
    <head>
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <title>7.5a2</title>
        <style>body { font: 16px "Segoe UI", "Tahoma", "Helvetica Neue", sans-serif }</style>
     </head>
     <body>
       <p>Behebt ein Problem im automatischen Update.</p>
    </body>
</html>
```

It seems IE automatically sets the content zoom to 200% in this setup.

<img width="513" alt="screen shot 2016-02-01 at 18 28 58" src="https://cloud.githubusercontent.com/assets/2419840/12725197/b1921a30-c911-11e5-977d-b9ec0972a2ed.png">

That's what my code change is supposed to do as well. But it turns out it is actually not working since the code is called too early. As a hack I tried to delay the call for some time and then it works. It seems it must be called from the document loaded event handler:

http://stackoverflow.com/questions/13377747/how-to-show-zoomed-out-web-browser-in-vb6-0

The "It doesn't seem to zoom text which specifies the font size in px." was actually a red herring. When I specify the font size in pt instead of px it works without code change.

So I have a working workaround. So feel free to close this bug or keep it open as low priority bug.

**Comment by staticfinal** [comment]
I would really like to provide you a full patch for this issue. Is there a reason why you use IWebBrowser directly instead of wxWebView? This would make it a lot easier to react to the document loaded event.

**Comment by vslavik** [comment]
> I would like to provide you with a fully working patch but I'm not familiar with COM programming in wxWidgets.

I'm sorry, but that's the laziest of lazy excuses. You're a *developer*, learning new things is what you *do*. I didn't know wx at first either, then I became its developer, then its user. I still don't know COM, that doesn't stop me from RTFMing and doing what I need to get done. If I had this attitude, you'd have no WinSparkle to use!

> Here are the release notes

Again, at least *complete* reproduction instructions *against the included sample* would be useful: i.e. a small patch to reproduce this in the sample. To rule out issues with your code and have something useful to work with. 

> `<meta http-equiv="X-UA-Compatible" content="IE=edge">`

More likely then not, there’s your issue — you’re telling the browser control that you have the renderer under control, but you really don’t. As mentioned above, `WebBrowser` defaults to compatibility mode, so explicitly turning compatibility mode off doesn’t seem to be a good idea (until this bug is closed, then it shouldn’t matter).

See https://poedit.net/updates/win/appcast for an example of a feed that doesn’t have the issues your super-secret feed has.

**Comment by vslavik** [comment]
> Is there a reason why you use IWebBrowser directly instead of wxWebView?

No. It’s historical, basically — WinSparkle code predates wxWebView. I suspect wxWebView would make `WinSparkle.dll` a bit larger, but that’s OK. I’m not sure if there’s enough user-code access to wxWebView to mess with `DOCHOSTUIFLAG_DPI_AWARE`, but if not, it’s certainly a fix that should go directly into wx (and I can commit it there), so that’s OK too. 

As a matter of fact, I have the switch to wxWebView on my TODO list for another reason: the ugly 3D border WinSparkle currently uses (not in your screenshot, though - is that `X-UA-Compatible` related?). wxWebView already contains the code necessary to remove the border.

In other words, I 100% agree: wxWebView is the way to go.

**Comment by staticfinal** [comment]
1. I'm developing my feed right now. So it is not yet deployed on a public server yet. Nothing super secret about it.
2. No the `<meta http-equiv="X-UA-Compatible" content="IE=edge">` not the issue. It only changes the border but not the zooming behavior.

But your attitude doesn't really motivate me to continue further with this investigation. Since your code is already perfect you probably shouldn't allow contributions nor bug reports.

**Comment by vslavik** [comment]
> But your attitude doesn't really motivate me

Wow. I wonder if you're referring to a) providing you with WinSparkle or b) trying to get to the bottom of this imperfect report despite being sabotaged on every step or c) trying to help with technical details or d) having way too much patience with your refusal to at the very least provide a reproducible case… Nope, the only "attitude" problem here is your entitlement and unwillingness to cooperate. If you did a better job with this issue, I wouldn't have to waste time asking you to fix its deficiencies, so don't talk to me about *motivation*.

> Since your code is already perfect

Do you realize how crazy this sounds? As you may recall, I described at least two defects above, but sure, don't let facts get in the way. You, on the other hand, *claim* there's a bug, but when I ask you to provide reproduction steps *so that I can fix it* (because you're clearly unwilling to help yourself), you repeatedly refuse, or are incapable of doing so. Any competent developer understands that reproducing a bug is the first and necessary step in fixing it and any reasonable developer understands that I can't magically clone *their* environment on my own. Surely you're both, so why you're so intent on sabotaging any chance of fixing this issue is beyond me.

Stop acting like you're doing me a favor and so damn entitled and get some perspective. It's I and other WinSparkle contributors that did something for *you* and you're only being asked to give a little back in the spirit of Open Source — a good bugreport, if not a PR. That shouldn't be so hard.

**Comment by wxiaoguang** [comment]
I am using the WebBrowser ActiveX in my wxWidgets app and met the same problem. At last, using the latest wxWidgets-3.1.0 resolved the High DPI problem. 
ps: I am not using WinSparkle so I am not sure whether it's the same problem.

wxWidgets 3.1.0 changes: https://github.com/wxWidgets/wxWidgets/blob/v3.1.0/docs/changes.txt
3.1.0: (released 2016-02-29)
----------------------------
- Many improvements for high DPI monitors support, notably XRC now interprets
  all pixel values as being in resolution-independent pixels.

**Comment by ThomasBarnekow** [comment]
@staticfinal I had a different issue with the `WebBrowser` control used by WinSparkle. That `WebBrowser` control renders HTML pages in Internet Explorer 7 (IE7, launched in 2006) mode by default.

My issue was solved by correctly configuring the `WebBrowser` control. Have a look at this [blog](https://weblog.west-wind.com/posts/2011/May/21/Web-Browser-Control-Specifying-the-IE-Version), for example, to see how that can be done.

**Comment by vslavik** [comment]
@ThomasBarnekow Kindly refrain from offtopic spamming of issues.

**Comment by vslavik** [comment]
The browser rendering code was significantly upgraded with f00e5c1af80f0b30cec6e9ce002b4a8ceca1d1ba and should fix this too (even with MSIE, but uses Edge now on any modern Windows system).
